Adição do Documento de Qualidade e Arquiteura. (#63)
fga-eps-mds/2022-1-Visualeasy-Doc: docs/documentacao/arquitetura.md

#   Arquitetura do projeto

### Finalidade

> <p align = "justify">O documento tem como objetivo estabelecer uma visão da arquitetura de software adotada no sistema Visualeasy. Com isso ele tem como finalidade evidenciar com clareza as decisões arquiteturais que foram tomadas no projeto. Disponibilizando as informações essenciais para as pessoas desenvolvedoras e demais envolvidos a respeito da aplicação e das tecnologias que serão utilizadas no desenvolvimento do projeto.</p>

Reviewer: ```suggestion
<p align = "justify">O documento estabelecerá uma visão da arquitetura de software adotada no sistema Visualeasy. Com isso ele tem como finalidade evidenciar com clareza as decisões arquiteturais tomadas no projeto. Disponibilizando as informações essenciais para as pessoas desenvolvedoras e demais envolvidos a respeito da aplicação e das tecnologias que serão utilizadas no desenvolvimento do projeto.</p>
```

> <p align = "justify">O documento tem como objetivo estabelecer uma visão da arquitetura de software adotada no sistema Visualeasy. Com isso ele tem como finalidade evidenciar com clareza as decisões arquiteturais que foram tomadas no projeto. Disponibilizando as informações essenciais para as pessoas desenvolvedoras e demais envolvidos a respeito da aplicação e das tecnologias que serão utilizadas no desenvolvimento do projeto.</p>

Reviewer 2: Colocar "software" em itálico

### Escopo

> <p align = "justify"> Essa documentação foi elaborada sobre a visão de arquitetura de software que será utilizada para a implementação do projeto Visualeasy, de forma a evidenciar as decisões tomadas. Serão abordados os padrões de arquitetura, os frameworks que serão usados no desenvolvimento do projeto. O objetivo é fornecer visualização gráfica de séries temporais, proporcionando uma melhor visualização dos dados fornecidos à aplicação e permitindo uma tomada de decisão com base na análise dos dados.</p>

Reviewer: ```suggestion
<p align = "justify"> Essa documentação foi elaborada sobre a visão de arquitetura de software que será utilizada para a implementação do projeto Visualeasy, para evidenciar as decisões tomadas. Serão abordados os padrões de arquitetura, os frameworks que serão usados no desenvolvimento do projeto. O objetivo é fornecer visualização gráfica de séries temporais, proporcionando uma melhor visualização dos dados fornecidos à aplicação e permitindo uma tomada de decisão com base na análise dos dados.</p>
```

> <p align = "justify"> Essa documentação foi elaborada sobre a visão de arquitetura de software que será utilizada para a implementação do projeto Visualeasy, de forma a evidenciar as decisões tomadas. Serão abordados os padrões de arquitetura, os frameworks que serão usados no desenvolvimento do projeto. O objetivo é fornecer visualização gráfica de séries temporais, proporcionando uma melhor visualização dos dados fornecidos à aplicação e permitindo uma tomada de decisão com base na análise dos dados.</p>

Reviewer 2: Mencionar as questões de qualidade que serão levadas em consideração durante o desenvolvimento; Colocar "framework" em itálico.

### Visão Geral

<p align = "justify">Aqui serão detalhados as características primordiais da arquitetura adotada pela equipe desenvolvedora. Estarão presentes a representação arquitetural, as restrições de arquitetura, visão lógica, de implementação e de dados.</p>

#### Representação da Arquitetura

> <p align = "justify">Para o desenvolvimento do sistema Visualeasy, foi decidido a utilização de uma arquitetura em microsserviços. Microsserviços são uma abordagem arquitetônica e organizacional do desenvolvimento de software na qual o software consiste em pequenos serviços independentes que se comunicam usando APIs bem definidas. Esses serviços pertencem a pequenas equipes autossuficientes.</p>

Reviewer: ```suggestion
<p align = "justify">Para o desenvolvimento do sistema Visualeasy, foi decidido a utilização de uma arquitetura em microsserviços. Microsserviços são uma abordagem arquitetônica e organizacional do desenvolvimento de software onde o software consiste em pequenos serviços independentes que se comunicam usando APIs bem definidas. Esses serviços pertencem a pequenas equipes autossuficientes.</p>
```

<p align ="justify">As arquiteturas de microsserviços facilitam a escalabilidade e agilizam o desenvolvimento de aplicativos, habilitando a inovação e acelerando o tempo de introdução de novos recursos no mercado.</p>

No sistema Visualeasy foi decidido a criação de 3 microsserviços, sendo eles:

- **Frontend**: Responsável pela visualização de todo o sistema.
- **Controle**: Responsável pelo envio e tratamento dos dados de todas as variáveis que serão visualizadas nos gráficos.
- **Autenticação**: Responsável por lidar com a autenticação no sistema e com a criação dos perfis de usuário dentro do sistema.

<p align ="justify">A utilização de microsserviços traz como benefícios o desacoplamento das funcionalidades do sistema. Portanto, a alteração de um deles ou até mesmo a queda de algum deles, não afetará o resto do sistema, podendo as outras funcionalidades continuarem em execução normalmente.</p>

<p align ="justify">Para o desenvolvimento dos microsserviços foi utilizado as seguintes tecnologias:</p>

#### Node JS

[![node](images/node.png)](images/node.png)

> <p align ="justify">Para o microsserviço de controle e de autenticação foi utilizado o Node.JS. Node é uma plataforma de aplicação, na qual os programas são escritos em JS e compilados, otimizados e interpretados pela Máquina Virtual V8, que é a mesma usada pelo Google para executar JavaScript no navegador Chrome. O resultado deste processo é entregue como código de máquina server-side, tornando o Node muito mais eficiente na sua execução e consumo de recursos. A plataforma Node foi escolhida por ser eficiente na implementação de aplicações em tempo real, que precisam transferir mensagens de um lado para o outro de forma rápida. As mensagens geradas pelo Visualeasy podem demandar um grande volume de dados a cada requisição do sistema.</p>
> 
> #### Next.JS
> 
> [![next](images/next.png)](images/next.png)
> 
> <p align ="justify">Para o microsserviço responsável pelo frontend foi utilizado o NextJS. O Next.JS é um framework para React que é uma biblioteca javascript, tendo seu foco no alto desempenho onde reúne diversas funcionalidades como renderização híbrida e estática de conteúdo, suporte a TypeScript, pre-fetching, sistema de rotas, pacotes de funcionalidades e diversos plugins, o framework será utilizado, como parte front-end do nosso projeto, em que mostrará as telas e apresentará os recursos de gráfico, login dentre outras funcionalidades.</p>
>    
> #### D3.js
> 
> [![d3](images/d3.png)](images/d3.png)
> <p align ="justify">Para a exibição dos gráficos foi escolhida a biblioteca D3.js. A D3.js é uma biblioteca em Javascript orientada a dados. Não é um framework monolítico, o que fornece uma melhor escalabilidade. D3 é rápido e trabalha com altos volumes de dados e interações com animações, assim evitando uma sobrecarga no sistema como um todo.</p>
> 
> #### PostgreSQL
> 
> [![postgres](images/postgres.png)](images/postgres.png)
> <p align ="justify">Para banco de dados foi escolhido o PostgreSQL. O PostgreSQL é um Sistema Gerenciador de Banco de Dados Relacional de código aberto que utiliza a linguagem SQL para armazenar de forma segura, extensível e possui um vasto ecossistema de ferramentas. É um software projetado para possuir compatibilidade com os principais sistemas operacionais como Linux e Windows. Por essas características esse Banco de Dados tem sido bastante utilizado no contexto geral de negócios, sites e por isso foi o escolhido como a solução de banco de dados do Visualeasy.</p>

Reviewer: ```suggestion
<p align ="justify">Para o microsserviço de controle e de autenticação foi utilizado o Node.JS. Node é uma plataforma de aplicação, onde os programas são escritos em JS e compilados, otimizados e interpretados pela Máquina Virtual V8, sendo a mesma usada pelo Google para executar JavaScript no navegador Chrome. O resultado deste processo é entregue como código de máquina server-side, tornando o Node muito mais eficiente na sua execução e consumo de recursos. A plataforma Node foi escolhida por ser eficiente na implementação de aplicações em tempo real, que precisam transferir mensagens de um lado para o outro de forma rápida. As mensagens geradas pelo Visualeasy podem demandar um grande volume de dados a cada requisição do sistema.</p>

#### Next.JS

[![next](images/next.png)](images/next.png)

<p align ="justify">Para o microsserviço responsável pelo front-end foi utilizado o NextJS. O Next.JS é um framework para React que é uma biblioteca javascript. O seu foco é no alto desempenho onde reúne diversas funcionalidades como renderização híbrida e estática de conteúdo. Possui suporte a TypeScript, pre-fetching, sistema de rotas, pacotes de funcionalidades e diversos plugins, o framework será utilizado, como parte front-end do nosso projeto, em que mostrará as telas e apresentará os recursos de gráfico, login dentre outras funcionalidades.</p>
   
#### D3.js

[![d3](images/d3.png)](images/d3.png)
<p align ="justify">Para a exibição dos gráficos foi escolhida a biblioteca D3.js. A D3.js é uma biblioteca em Javascript orientada a dados. Não é um framework monolítico, fornecendo uma melhor escalabilidade. D3 é rápido e trabalha com altos volumes de dados e interações com animações, assim evitando uma sobrecarga no sistema todo.</p>

#### PostgreSQL

[![postgres](images/postgres.png)](images/postgres.png)
<p align ="justify">Para banco de dados foi escolhido o PostgreSQL. O PostgreSQL é um Sistema Gerenciador de Banco de Dados Relacional de código aberto que utiliza a linguagem SQL para armazenar seguramente os dados, extensível e possui um vasto ecossistema de ferramentas. É um software projetado para possuir compatibilidade com os principais sistemas operacionais como Linux e Windows. Por essas características esse Banco de Dados tem sido bastante utilizado no contexto geral de negócios, sites por isso foi o escolhido como a solução de banco de dados do Visualeasy.</p>
```

|Virtualização|Docker|20.10.17|Plataforma aberta para empacotamento e execução de aplicações em contêineres|
|Virtualização|Docker-compose|3.9|Gerenciamento de contêiner|
|Biblioteca|Jest|28.1.3|Biblioteca de testes|

### Visão Lógica
Detalhar micro serviços aqui

### Arquitetura dos Serviços e visão de Implementação

#### Visão Geral

[![estrutura](images/estrutura.png)](images/estrutura.png)

Cada microsserviço é detalhado da seguinte forma:

- <p align ="justify">Visualização de dados/gráficos : Front-end do produto composto pelo framework Next.Js, para a construção das telas, campos, validações e botões que o usuário irá interagir. Acompanhando o Next.Js, o D3.Js é responsável pela visualização dos dados tratados, gerando os gráficos.</p>
- <p align ="justify">O Core é composto pelas APIs responsáveis pela autenticação de usuário e pelo controle, coleta de dados e geração de gráficos.</p>

## Visão de Dados

## Qualidade

<p align ="justify">Para garantir a qualidade do sistema Visualeasy, serão seguidas 6 características de qualidade de software, definidas pela ISO9126 (em português, NBR13596). Cada característica está detalhada a seguir.</p>

### Funcionalidade

> <p align ="justify">A funcionalidade de um software diz respeito à satisfação de necessidades que deram origem ao projeto. Abrange requisitos implícitos e explícitos e está intimamente ligada à qualidade do código criado. Neste sentido, o Visualeasy deve estar de acordo com o objetivo que originou sua demanda, e deve atender as necessidades que foram previstas para o usuário comum.</p>
> 
> <p align ="justify">A característica de funcionalidade abrange também questões de segurança do sistema. O desenvolvimento do Visualeasy deve garantir que apenas usuários autorizados tenham acesso às informações do sistema. Para isso, será implementado um sistema de autenticação de usuários, com a utilização de tokens para validação de acessos.</p>
> 
> ### Confiabilidade
> 
> <p align ="justify">É a garantia que o software dá em não apresentar falhas em um período de tempo e ambiente específico. O Visualeasy busca apresentar os dados fidedignos em gráficos de acordo com o requisitado no banco de dados.
> Usabilidade</p>
> 
> <p align ="justify">A usabilidade está relacionada à meta de garantir que os usuários usem bem as funcionalidades de um sistema. A navegabilidade do Visualeasy deve ser intuitiva e fácil de ser utilizada, para que os usuários possam acessar as funcionalidades disponíveis sem grande esforço.</p>

Reviewer: ```suggestion
<p align ="justify">A funcionalidade de um software diz respeito à satisfação de necessidades que deram origem ao projeto. Abrange requisitos implícitos e explícitos e está intimamente ligada à qualidade do código criado. Neste sentido, o Visualeasy deve corresponder ao objetivo que originou sua demanda, e deve atender as necessidades previstas para o usuário comum.</p>

<p align ="justify">A característica de funcionalidade abrange também questões de segurança do sistema. O desenvolvimento do Visualeasy deve garantir que apenas usuários autorizados tenham acesso às informações do sistema. Para isso, será implementado um sistema de autenticação de usuários, com a utilização de tokens para validação de acessos.</p>

### Confiabilidade

<p align ="justify">É a garantia que o software dá em não apresentar falhas em um período de tempo e ambiente específico. O Visualeasy busca apresentar os dados fidedignos em gráficos conforme o requisitado no banco de dados.
Usabilidade</p>

<p align ="justify">A usabilidade está relacionada à meta de garantir que os usuários usem bem as funcionalidades de um sistema. A navegabilidade do Visualeasy deve ser intuitiva e fácil de ser utilizada, para que os usuários possam acessar as funcionalidades disponíveis sem grande esforço.</p>
```

### Eficiência

> <p align ="justify">A eficiência se refere à forma como o sistema ajuda os usuários a realizar suas tarefas. O Visualeasy deve oferecer formas fáceis e rápidas para que o usuário encontre a funcionalidade ou informações que deseja. Além disso, o produto deve ser compatível com o nível que foi requerido para o produto. Sendo assim, o Visualeasy deve ser eficaz às necessidades relatadas pelos Product Owners.</p>

Reviewer: ```suggestion
<p align ="justify">A eficiência se refere à forma como o sistema ajuda os usuários a realizar suas tarefas. O Visualeasy deve oferecer formas fáceis e rápidas para que o usuário encontre a funcionalidade ou informações que deseja. Além disso, o produto deve ser compatível com o nível requerido para o produto. Sendo assim, o Visualeasy deve ser eficaz e atender às necessidades dos Product Owners.</p>
```

### Manutenibilidade

> <p align ="justify">A manutenibilidade está diretamente ligada à qualidade do software que produziremos. Manutenções corretivas devem ser constantes a fim de corrigir bugs ou outros problemas ligados a utilização do sistema. O sistema deve ser capaz de se adaptar às mudanças do ambiente no qual vai atuar. Por exemplo, atualizar o sistema operacional ou tecnologias, bibliotecas.</p>

Reviewer: ```suggestion
<p align ="justify">A manutenibilidade está diretamente ligada à qualidade do software que produziremos. Manutenções corretivas devem ser constantes a fim de corrigir bugs ou outros problemas ligados a utilização do sistema. O sistema deve ser capaz de se adaptar às mudanças do ambiente onde irá atuar. Por exemplo, atualizar o sistema operacional ou tecnologias, bibliotecas.</p>
```

* Usabilidade e suas metas. Irla Rebelo, IHC na prática. Disponível em: [Irlabr](https://irlabr.wordpress.com/apostila-de-ihc/parte-1-ihc-na-pratica/6-usabilidade-e-suas-metas). Acesso em 21 de Julho de 2022.
* Rogers et al. (2013). Design de Interação: Além da Interação Humano-Computador - Capítulo 1 - O que é design de interação.
* Barbosa, S. D. J. and Silva, B. S. d. (2010). Interação Humano-Computador. Rio de Janeiro: Elsevier.

## Versionamento

| Data | Versão | Descrição | Autor(es) |
|------|------|------|------|
|19/07/2022|1.0|Criação do documento de arquitetura do projeto| [Bruna Santos](https://github.com/brunaalmeidasantos),[Estevão Reis](https://github.com/estevaoreis25), [Itallo Gravina](https://github.com/itallogravina), [Luis Bruno](https://github.com/lbrunofidelis), [Damarcones Porto](https://github.com/damarcones), [Bruno Nunes](https://github.com/brunocmo), [Marcos Vinicius](https://github.com/marcos-mv), [Gustavo Moreira](https://github.com/gustavoduartemoreira), [Gabriel Batalha](https://github.com/gustavoduartemoreira), [João Pedro](https://github.com/Joao-Pedro-Moura)|

> |23/07/2022|2.0|Correções e Adição do documento ao reposítório | [Marcos Vinicius](https://github.com/marcos-mv)

Reviewer: ```suggestion
|23/07/2022|1.1|Correções e Adição do documento ao reposítório | [Marcos Vinicius](https://github.com/marcos-mv)
|25/07/2022|1.2|Revisão de ortográfica do documento|[Gustavo Moreira](https://github.com/gustavoduartemoreira)|
```
